File: app/src/main/java/alejandromr92/com/movies/presentation/presenter/impl/GetPopularMoviesPresenterImpl.java
package alejandromr92.com.movies.presentation.presenter.impl;

import alejandromr92.com.movies.domain.interactor.GetPopularMoviesInteractor;
import alejandromr92.com.movies.domain.interactor.base.Executor;
import alejandromr92.com.movies.domain.interactor.base.MainThread;
import alejandromr92.com.movies.domain.interactor.impl.GetPopularMoviesInteractorImpl;
import alejandromr92.com.movies.domain.model.Movie;
import alejandromr92.com.movies.presentation.model.converters.MovieViewconverter;
import alejandromr92.com.movies.presentation.presenter.BasePresenter;
import alejandromr92.com.movies.presentation.presenter.GetPopularMoviesPresenter;
import alejandromr92.com.movies.storage.repositories.MovieRepository;

import java.util.List;

public class GetPopularMoviesPresenterImpl extends BasePresenter implements GetPopularMoviesPresenter, GetPopularMoviesInteractor.Callback {

    private View view;

    private MovieRepository movieRepository;

    public GetPopularMoviesPresenterImpl(Executor executor, MainThread mainThread, MovieRepository movieRepository, View view) {
        super(executor, mainThread);
        this.view = view;
        this.movieRepository = movieRepository;
    }

    @Override
    public void getPopularMovies(int page) {
        GetPopularMoviesInteractor interactor = new GetPopularMoviesInteractorImpl(executor, mainThread, movieRepository, this);
        interactor.execute(page);
    }

    @Override
    public void onPopularMoviesRetrieved(List<Movie> movieList) {
        view.onPopularMoviesRetrieved(MovieViewconverter.convertListToViewModel(movieList));
    }

    @Override
    public void onPopularMoviesRetrievingError(int errorCode) {
        view.onPopularMoviesRetrievingError(errorCode);
    }
}

File: app/src/main/java/alejandromr92/com/movies/presentation/presenter/impl/GetPopularMoviesPresenterImpl.kt
package alejandromr92.com.movies.presentation.presenter.impl

import alejandromr92.com.movies.domain.interactor.impl.GetPopularMoviesInteractorImpl
import alejandromr92.com.movies.domain.model.Movie
import alejandromr92.com.movies.presentation.model.converters.MovieViewconverter
import alejandromr92.com.movies.presentation.presenter.GetPopularMoviesPresenter
import alejandromr92.com.movies.storage.repositories.MovieRepository
import io.reactivex.Scheduler

class GetPopularMoviesPresenterImpl(
    private val threadExecutor: Scheduler,
    private val mainThread: Scheduler,
    private val view: GetPopularMoviesPresenter.View,
    private val movieRepository: MovieRepository
) : GetPopularMoviesPresenter{

    override fun getPopularMovies(page: Int) {
        view.showProgress()
        val interactor = GetPopularMoviesInteractorImpl(movieRepository, mainThread, threadExecutor)
        interactor.execute(page, ::onPopularMoviesRetrieved, ::onPopularMoviesRetrievingError)
    }

    private fun onPopularMoviesRetrieved(movieList: List<Movie>){
        view.hideProgress()
        view.onPopularMoviesRetrieved(MovieViewconverter.convertListToViewModel(movieList))
    }

    private fun onPopularMoviesRetrievingError(throwable: Throwable){
        view.hideProgress()
        view.onPopularMoviesRetrievingError()
    }

}